ledger/complete/mtrie/node/node.go

```diff
@@ -160,83 +131,115 @@ func NewInterimNode(height int, lchild, rchild *Node) *Node {
 	}
 
 	// this is an INTERIOR node at least one of lChild or rChild is not nil.
-	var h1, h2 []byte
+	var h1, h2 hash.Hash
 	if n.lChild != nil {
 		h1 = n.lChild.Hash()
 	} else {
-		h1 = common.GetDefaultHashForHeight(n.height - 1)
+		h1 = hash.GetDefaultHashForHeight(n.height - 1)
 	}
 
 	if n.rChild != nil {
 		h2 = n.rChild.Hash()
 	} else {
-		h2 = common.GetDefaultHashForHeight(n.height - 1)
+		h2 = hash.GetDefaultHashForHeight(n.height - 1)
 	}
-	common.HashInterNodeIn(result, h1, h2)
+	return hash.HashInterNodeIn(h1, h2)
 }
 
 // VerifyCachedHash verifies the hash of a node is valid
-func (n *Node) verifyCachedHashRecursive(computedHash *[]byte) bool {
-	if n.lChild != nil {
-		if !n.lChild.verifyCachedHashRecursive(computedHash) {
-			return false
-		}
+func verifyCachedHashRecursive(n *Node) bool {
+	if n == nil {
+		return true
 	}
-	if n.rChild != nil {
-		if !n.rChild.verifyCachedHashRecursive(computedHash) {
-			return false
-		}
+	if !verifyCachedHashRecursive(n.lChild) {
+		return false
 	}
 
-	if n.hashValue != nil {
-		n.computeHash(computedHash)
-		return bytes.Equal(n.hashValue, *computedHash)
+	if !verifyCachedHashRecursive(n.rChild) {
+		return false
 	}
-	return true
+
+	computedHash := n.computeHash()
+	return n.hashValue == computedHash
 }
 
 // VerifyCachedHash verifies the hash of a node is valid
 func (n *Node) VerifyCachedHash() bool {
-	computedHash := make([]byte, common.HashLen)
-	return n.verifyCachedHashRecursive(&computedHash)
+	return verifyCachedHashRecursive(n)
 }
 
 // Hash returns the Node's hash value.
 // Do NOT MODIFY returned slice!
-func (n *Node) Hash() []byte { return n.hashValue }
+func (n *Node) Hash() hash.Hash {
+	if n == nil {
+		// case of an empty trie root
+		return hash.GetDefaultHashForHeight(hash.TreeMaxHeight)
+	}
+	return n.hashValue
+}
 
 // Height returns the Node's height.
 // Per definition, the height of a node v in a tree is the number
 // of edges on the longest downward path between v and a tree leaf.
-func (n *Node) Height() int { return n.height }
+func (n *Node) Height() int {
+	if n == nil {
+		return hash.TreeMaxHeight
+	}
+	return n.height
+}
 
 // MaxDepth returns the longest path from this node to compacted leafs in the subtree.
 // in contrast to the Height, this value captures compactness of the subtrie.
-func (n *Node) MaxDepth() uint16 { return n.maxDepth }
+func (n *Node) MaxDepth() uint16 {
+	if n == nil {
+		return 0
+	}
+	return n.maxDepth
+}
 
 // RegCount returns number of registers allocated in the subtrie of this node.
-func (n *Node) RegCount() uint64 { return n.regCount }
+func (n *Node) RegCount() uint64 {
+	if n == nil {
+		return 0
+	}
+	return n.regCount
+}
 
 // Path returns the the Node's register storage path.
-func (n *Node) Path() ledger.Path { return n.path }
+//
+// the path of an nil node is defined as the arbitrary value 00..00, although
+// no node or trie algorithm is relying on this value.
+func (n *Node) Path() ledger.Path {
```

Comment:
⚠️ I am worried about this convention. Path `00..00` is still a valid path. Essentially, what you are doing is giving the nil instance exactly the same properties as a compactified leaf with empty payload. I feel then we might as well make this explicit and add the actual compactified leaf.